### PR TITLE
[SPARK-33477][SQL] Hive Metastore support filter by date type

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

Hive Metastore supports strings and integral types in filters. It could also support dates. Please see [HIVE-5679](https://github.com/apache/hive/commit/5106bf1c8671740099fca8e1a7d4b37afe97137f) for more details.

This pr add support it.

### Why are the changes needed?

Improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 1]
```diff
@@ -1265,11 +1265,13 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
       defaultTimeZoneId: String): Seq[CatalogTablePartition] = withClient {
     val rawTable = getRawTable(db, table)
     val catalogTable = restoreTableMetadata(rawTable)
+    val timeZoneId = CaseInsensitiveMap(catalogTable.storage.properties).getOrElse(
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:1268]: Why does `prunePartitionsByFilter` not use it?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 2]
```diff
@@ -701,6 +710,21 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
       }
     }
 
+    object ExtractableDateValues {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:713]: why not update `ExtractableValues` to support date？

Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:713]: This is because for `InSet` predicate, the `hset`'s value is the int type, we need to use `dateFormatter` to format it, this is the test:
https://github.com/apache/spark/blob/ba2f5531601ff169ff0f9fab9bb60627f65082bf/sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HivePartitionFilteringSuite.scala#L339-L342

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala (hunk 1 above)
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:1268]: `prunePartitionsByFilter` also use it:
https://github.com/apache/spark/blob/0032d85153e34b9ac69598b7dff530094ed0f640/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala#L157

https://github.com/apache/spark/blob/dfa6fb46f4238792bff6a0201da201be1b42620e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala#L152-L163

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/FiltersSuite.scala [hunk 3]
```diff
@@ -63,6 +64,24 @@ class FiltersSuite extends SparkFunSuite with Logging with PlanTest {
     (Literal(1) === a("intcol", IntegerType)) :: (Literal("a") === a("strcol", IntegerType)) :: Nil,
     "1 = intcol and \"a\" = strcol")
 
+  filterTest("date filter",
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/FiltersSuite.scala:67]: do we run these test with different hive versions?

Comment (wangyum) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/client/FiltersSuite.scala:67]: Different hive versions tested by `HivePartitionFilteringSuite`:
https://github.com/apache/spark/blob/ba2f5531601ff169ff0f9fab9bb60627f65082bf/sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HivePartitionFilteringSuite.scala#L290-L343

**Comment by cloud-fan** [comment]
@wangyum can you resolve the conflicts? thanks!

**Comment by cloud-fan** [comment]
last question about correctness: Does hive execute the partition predicate as date comparison or string comparison? The later can be problematic.

**Comment by wangyum** [comment]
It's date comparison: https://github.com/apache/hive/blob/rel/release-2.3.7/metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java#L1141-L1148

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/FiltersSuite.scala [hunk 4]
```diff
@@ -63,6 +64,24 @@ class FiltersSuite extends SparkFunSuite with Logging with PlanTest {
...
+      (Literal(Date.valueOf("2019-01-01")), Literal(Date.valueOf("2019-01-07")))) :: Nil,
+    "(datecol = 2019-01-01 or datecol = 2019-01-07)")
+
+  filterTest("date and string filter",
+    (Literal(Date.valueOf("2019-01-01")) === a("datecol", DateType)) ::
+      (Literal("a") === a("strcol", IntegerType)) :: Nil,
+    "2019-01-01 = datecol and \"a\" = strcol")
+
+  filterTest("date filter with null",
+    (a("datecol", DateType) ===  Literal(null)) :: Nil,
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/FiltersSuite.scala:82]: not related to this PR, but we can pushdown `col is null` predicate to hive for this case.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HivePartitionFilteringSuite.scala [hunk 5]
```diff
@@ -297,6 +294,61 @@ class HivePartitionFilteringSuite(version: String)
...
+          } yield Set(
+            "part" -> dt.toString
+          )
+      }.reduce(_ ++ _)
+
+      assert(filteredPartitions.map(_.spec.toSet).toSet == expectedPartitions.toSet)
+    }
+
+    testDataTypeFiltering(
+      Seq(AttributeReference("part", DateType)() === Date.valueOf("2019-01-01")),
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HivePartitionFilteringSuite.scala:337]: can we create an `attr` method to get the `AttributeReference` from the table? to follow other tests.

**Review by cloud-fan — APPROVED** [review]
LGTM except one comment for test

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 6]
```diff
@@ -700,6 +709,21 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
...
+      private lazy val valueToLiteralString: PartialFunction[Any, String] = {
+        case value: Int => dateFormatter.format(value)
+      }
+
+      def unapply(values: Set[Any]): Option[Seq[String]] = {
+        val extractables = values.toSeq.map(valueToLiteralString.lift)
+        if (extractables.nonEmpty && extractables.forall(_.isDefined)) {
```
Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:719]: Why do we need `forall` here? `InSet` can have mixed values: int and other types?

**Review by maropu — APPROVED** [review]
The other parts looks fine.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala (hunk 6 above)
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:719]: Otherwise this test will fail:
```scala
  filterTest("string filter with InSet predicate",
    (InSet(a("stringcol", StringType),
      Range(1, 10).map(d => UTF8String.fromString(d.toString)).toSet)) :: Nil,
    "")
```

```
None.get
java.util.NoSuchElementException: None.get
	at scala.None$.get(Option.scala:529)
	at scala.None$.get(Option.scala:527)
	at org.apache.spark.sql.hive.client.Shim_v0_13$ExtractableDateValues$1$.$anonfun$unapply$7(HiveShim.scala:720)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
	at scala.collection.Iterator.foreach(Iterator.scala:941)
	at scala.collection.Iterator.foreach$(Iterator.scala:941)
```

Comment (maropu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:719]: Ah, ok. Thanks.

**Comment by cloud-fan** [comment]
retest this please

**Comment by wangyum** [comment]
@shaneknapp Did you set :`export LANG=en_US.UTF-8`?
```
org.apache.spark.sql.AnalysisException: org.apache.hadoop.hive.ql.metadata.HiveException: MetaException(message:java.nio.file.InvalidPathException: Malformed input or input contains unmappable characters: /home/jenkins/workspace/SparkPullRequestBuilder@3/sql/hive/target/tmp/hive_execution_test_group/warehouse-1355e680-268f-4224-b549-eaddcadcf136/DaTaBaSe_I.db/tab_ı);
```

This issue should be fixed if we set `export LANG=en_US.UTF-8`, more details:https://issues.apache.org/jira/browse/SPARK-27177

**Comment by maropu** [comment]
@wangyum How about asking it in the spark-dev thread so that Shane could notice it quickly?http://apache-spark-developers-list.1001551.n3.nabble.com/jenkins-downtime-tomorrow-evening-weekend-tt30405.html

**Comment by wangyum** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 7]
```diff
@@ -711,7 +735,8 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
         val resolver = SQLConf.get.resolver
         if (varcharKeys.exists(c => resolver(c, attr.name))) {
           None
-        } else if (attr.dataType.isInstanceOf[IntegralType] || attr.dataType == StringType) {
+        } else if (attr.dataType.isInstanceOf[IntegralType] || attr.dataType == StringType ||
+          attr.dataType == DateType) {
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:739]: indentation?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 8]
```diff
@@ -748,6 +773,10 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
         convert(And(GreaterThanOrEqual(child, Literal(sortedValues.head, dataType)),
           LessThanOrEqual(child, Literal(sortedValues.last, dataType))))
 
+      case InSet(child @ ExtractAttribute(SupportedAttribute(name)), ExtractableDateValues(values))
+        if useAdvanced && child.dataType == DateType =>
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:777]: indentation?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/FiltersSuite.scala [hunk 9]
```diff
@@ -63,6 +65,28 @@ class FiltersSuite extends SparkFunSuite with Logging with PlanTest {
...
+    "")
+
+  filterTest("string filter with InSet predicate",
+    InSet(a("strcol", StringType), Set("1", "2").map(s => UTF8String.fromString(s))) :: Nil,
+    "(strcol = \"1\" or strcol = \"2\")")
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/FiltersSuite.scala:88]: nit. This seems to be irrelevant to this PR, but it's good to have this.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/FiltersSuite.scala [hunk 10]
```diff
@@ -139,6 +163,11 @@ class FiltersSuite extends SparkFunSuite with Logging with PlanTest {
         InSet(a("doublecol", DoubleType),
           Range(1, 20).map(s => Literal(s.toDouble).eval(EmptyRow)).toSet),
         "")
+
+      checkConverted(
+        InSet(a("datecol", DateType),
+          Range(1, 20).map(d => Literal(d, DateType).eval(EmptyRow)).toSet),
+        "(datecol >= 1970-01-02 and datecol <= 1970-01-20)")
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/FiltersSuite.scala:170]: Oh, is this not `datecol >= 1970-01-20 and ...`?

Comment (dongjoon-hyun) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/client/FiltersSuite.scala:170]: Never mind. I found the reason.

**Review by dongjoon-hyun — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/FiltersSuite.scala (hunk 10 above)
Comment (wangyum) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/client/FiltersSuite.scala:170]:
```scala
scala> import org.apache.spark.sql.types._
import org.apache.spark.sql.types._

scala> import org.apache.spark.sql.catalyst.expressions._
import org.apache.spark.sql.catalyst.expressions._

scala> Range(1, 20).map(d => Cast(Literal(d, DateType), StringType, Some("America/Los_Angeles"))).map(_.eval(EmptyRow))
res13: scala.collection.immutable.IndexedSeq[Any] = Vector(1970-01-02, 1970-01-03, 1970-01-04, 1970-01-05, 1970-01-06, 1970-01-07, 1970-01-08, 1970-01-09, 1970-01-10, 1970-01-11, 1970-01-12, 1970-01-13, 1970-01-14, 1970-01-15, 1970-01-16, 1970-01-17, 1970-01-18, 1970-01-19, 1970-01-20)
```

**Comment by wangyum** [comment]
retest this please.

**Comment by HyukjinKwon** [comment]
Merged to master.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HivePartitionFilteringSuite.scala [hunk 11]
```diff
@@ -297,6 +294,63 @@ class HivePartitionFilteringSuite(version: String)
...
+      identifier = TableIdentifier("test_date", Some("default")),
+      tableType = CatalogTableType.MANAGED,
+      schema = new StructType().add("value", "int").add("part", "date"),
+      partitionColumnNames = Seq("part"),
+      storage = storageFormat)
+    client.createTable(table, ignoreIfExists = false)
+
+    val partitions =
+      for {
+        date <- Seq("2019-01-01", "2019-01-02", "2019-01-03", "2019-01-04")
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/HivePartitionFilteringSuite.scala:308]: How about NULL?

**Comment by gatorsmile** [comment]
@wangyum Could you add more test cases to check the NULL handling cases? For example, 
- Include NULL values in the data set
- Include NULL values in the predicates
- Include null-safe equals 

Please check https://spark.apache.org/docs/3.0.1/sql-ref-null-semantics.html#comp-operators

**Comment by wangyum** [comment]
> @wangyum Could you add more test cases to check the NULL handling cases? For example,
> 
> * Include NULL values in the data set
> * Include NULL values in the predicates
> * Include null-safe equals
> 
> Please check https://spark.apache.org/docs/3.0.1/sql-ref-null-semantics.html#comp-operators

OK
